### PR TITLE
TCES-13 Client Backend APIs

### PR DESCRIPTION
**Pull request by VangoCode**
Creates backend APIs for Client CRUD operations. Also implements mock creation for unit tests.

Completes issue [TCES-13](https://linear.app/uoftblueprint/issue/TCES-13/clt-client-backend-apis)

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist

Please delete options that are not applicable.

- [x] My code follows the style guidelines of this project (required)
- [x] I have performed a self-review of my code (required)
- [x] I have linted and formatted my code (required)
- [x] New and existing unit tests pass locally with my changes (required)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: server/src/controllers/client/addClients.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    var closure_date = null;
+    if (req.body.client.closure_date) {
+      closure_date = new Date(req.body.client.closure_date);
+    }
+
+    // create one client
+    const client = await Client.create({
+      owner: req.body.client.owner,
+      creator: req.body.client.creator,
+      name: req.body.client.name || null,
```
Comment (DanielDervishi) [inline server/src/controllers/client/addClients.js]: Name will never be an empty field

File: server/src/models/client.model.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      isInUser(value) {
+        if (!User.findByPk(value)) {
+          throw new Error("User does not exist");
+        }
+      },
+    },
+  },
+  name: {
+    type: DataTypes.STRING,
+    allowNull: true,
```
Comment (DanielDervishi) [inline server/src/models/client.model.js]: change to false

File: server/src/controllers/client/addClients.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      email: req.body.client.email || null,
+      phone_number: req.body.client.phone_number || null,
+      status: req.body.client.status || "active",
+      closure_date: closure_date || null,
+      status_at_exit: "no_results",
+      status_at_3_months: "no_results",
+      status_at_6_months: "no_results",
+      status_at_12_months: "no_results",
```
Comment (DanielDervishi) [inline server/src/controllers/client/addClients.js]: These should probably be null on creation

File: server/src/controllers/client/updateClient.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
+const logger = require("pino")();
+const Client = require("../../models/client.model");
+
+const updateClientRequestHandler = async (req, res) => {
+  try {
+    const client_id = req.params.client_id;
+    const client = await Client.findOne({ where: { id: client_id } });
+    if (!client) {
```
Comment (DanielDervishi) [inline server/src/controllers/client/updateClient.js:9]: When updating the client's status to active, we probably want to clear status_at_exit, status_at_x_months, as well as closure date

File: server/src/controllers/client/updateClient.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
+const logger = require("pino")();
+const Client = require("../../models/client.model");
+
+const updateClientRequestHandler = async (req, res) => {
```
Comment (DanielDervishi) [inline server/src/controllers/client/updateClient.js:4]: Given the recent model changes we should update the date_updated attribute of the client model

File: server/src/controllers/client/addClients.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
+const logger = require("pino")();
+const Client = require("../../models/client.model");
+
+const addClientsRequestHandler = async (req, res) => {
```
Comment (DanielDervishi) [inline server/src/controllers/client/addClients.js:4]: Given the recent model changes, we should be adding the date_created to each of the models.(Sequelize may provide the ability to auto set this attribute by adding something to the model definition)

File: server/src/controllers/client/updateClient.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+          status: "fail",
+          message: "You must provide a status_at_exit when closing a client",
+          data: null,
+        });
+      }
+
+      req.body.values.closed_date = new Date();
+    }
+
+    await client.set(req.body.values);
```
Comment (DanielDervishi) [inline server/src/controllers/client/updateClient.js:52]: closure date should probably be auto set when the status is being changed to closed to the current date and time (given that the status is not already closed)

File: server/src/routes/client.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ */
+router.post("/", isLoggedIn, addClientsRequestHandler);
+
+/**
+ * Update a specific client's info, with id client_id
+ *
+ * Expected parameters:
+ * @type string {params.client_id}
+ * Expected body parameters:
+ * @type Client {params.body.client}
```
Comment (DanielDervishi) [inline server/src/routes/client.js]: This should probably be params.body.values in compliance with updateClientRequestHandler

File: server/src/controllers/client/addClients.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+    var closure_date = null;
+    if (req.body.client.closure_date) {
+      closure_date = new Date(req.body.client.closure_date);
+    }
+
+    // create one client
+    const client = await Client.create({
+      owner: req.body.client.owner,
+      creator: req.body.client.creator,
```
Comment (DanielDervishi) [inline server/src/controllers/client/addClients.js]: these should be automatically set (they should refer to the currently logged in user)

**Review by DanielDervishi — COMMENTED** [review]
Great work Ron!

File: server/src/models/client.model.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
+require("dotenv").config();
```
Comment (jordanjanakievski) [inline server/src/models/client.model.js:1]: More so general comment reminder for updating some fields/adding new ones. I know Daniel has a few comments related to date_added.

**Review by jordanjanakievski — COMMENTED** [review]
Looking good. Daniel's comments have most of the stuff covered.
If you are happy with your test coverage then I am also happy with your test coverage

Screenshot 2023-11-21 at 00 43 20

Add Clients might be the only one I would suggest taking an extra look at since its branches are failing under 50%. But since it's your code I know you will know what's best.

**Review by DanielDervishi — APPROVED** [review]
Great Work Ron!

**Review by jordanjanakievski — APPROVED** [review]
Great work! Thanks for getting this done and the tests all look great too!

**Pull request merged**